File: 2022/c/01_calories/calories2.c
// elf calories
#include <stdio.h>
#include <stdlib.h>

#define SIZE 10

int main() {
    FILE* file = NULL;
    int calories[3] = { 0 }, elves[3] = { 0 }, elfCount = 1;
    char elfCurrent[SIZE]; // char for fgets

    if (!(file = fopen("input.txt", "r"))) {
        puts("FILE DID NOT OPEN");
        exit(EXIT_FAILURE);
    }

    while (fgets(elfCurrent, SIZE, file)) {
        int totalElf = 0;
        printf("\n Elf %i\n -------\n", elfCount);
        for (int i = 1; elfCurrent[0] != '\n' && !feof(file); i++) {
            printf("%2i: %s", i, elfCurrent);
            totalElf += atoi(elfCurrent);
            fgets(elfCurrent, SIZE, file);
        }
        printf("TOTAL: %i", totalElf);
        if (totalElf > calories[0]) {
            calories[2] = calories[1];
            calories[1] = calories[0];
            calories[0] = totalElf;
            elves[2] = elves[1];
            elves[1] = elves[0];
            elves[0] = elfCount; // biggest elf
        }
        elfCount++;
        puts("");
    }

    fclose(file);

    int totalElf = 0;
    for (size_t i = 0; i < 3; i++) {
        printf("\nElf %3i: %i", elves[i], calories[i]);
        totalElf += calories[i];
    }
    printf("\n\nElf %i has the most calories", elves[0]);
    printf("\nSum of 3 largest elves: %i", totalElf);
}

File: 2022/go/01_calories/calories2.go
package main

import (
    "bufio"
    "fmt"
    "os"
    "strconv"
)

func main() {
    elf_count := 1
    elves := make([]int, 3)
    calories := make([]int, 3)

    file, err := os.Open("input")
    if err != nil {
        fmt.Fprintf(os.Stderr, "%s", err)
    }
    defer file.Close()

    buf := bufio.NewScanner(file)

    elf := ""
    for buf.Scan() {
        fmt.Printf("\n Elf %d\n -------\n", elf_count)

        elf = buf.Text()
        if elf == "" {
            continue
        }

        total := 0
        for i := 1; elf != "" && buf.Scan(); i++ {
            calories, _ := strconv.Atoi(elf)
            fmt.Printf("%2d: %d\n", i, calories)
            total += calories
            elf = buf.Text()
        }

        fmt.Printf("total: %d\n", total)
        if total > calories[0] {
            calories[2] = calories[1]
            calories[1] = calories[0]
            calories[0] = total
            elves[2] = elves[1]
            elves[1] = elves[0]
            elves[0] = elf_count
        }
        elf_count++
    }
    fmt.Println("")

    total := 0
    for i, v := range calories {
        fmt.Printf("Elf %d: %d\n", elves[i], v)
        total += v
    }

    fmt.Printf("\nElf %d has them most calories: %d", elves[0], calories[0])
    fmt.Printf("\nSum of three largest elves: %d", total)
}

// https://kgrz.io/reading-files-in-go-an-overview.html
